core/src/ast/types.rs

```diff
@@ -379,6 +404,7 @@ pub enum TypeName {
     /// The path must be present! Ordering will be parsed as an AST type!
     Ordering,
     Function(Vec<Box<TypeName>>, Box<TypeName>),
+    TraitImpl(PathType),
```

Comment:
nit: ImplTrait or TraitObject

Reply:
also nit: similar to callbacks, this should only be a meaningful variant on input types.

Reply:
Good point! I'll add the `InputOnly` constraint on the HIR side, same as for Callbacks